webapp/certified/views.py

```diff
@@ -0,0 +1,353 @@
+        component=component,
+        machines=sorted(
+            machines, key=lambda machine: machine["canonical_id"], reverse=True
+        ),
+    )
+
+
+def certified_hardware_details(canonical_id, release):
+
+    try:
+        models = api.certified_models(canonical_id=canonical_id)["objects"][0]
+    except KeyError:
+        abort(404)
+
+
+    model_releases = api.certified_model_details(
+        canonical_id=canonical_id, limit="0"
+    )["objects"]
+
+    # Release section
+    release_details = next(
+        (
+            detail
+            for detail in model_releases
+            if detail["certified_release"] == release
+        ),
+        None,
+    )
+    if not release_details:
+        abort(404)
+
+    model_devices = api.certified_model_devices(
+        canonical_id=canonical_id, limit="0"
+    )["objects"]
+
+    hardware_details = {}
+    for device in model_devices:
+        device_info = {
+            "name": (
+                f"{device['make']} {device['name']}"
+                f" {device['subproduct_name']}"
+            ),
+            "bus": device["bus"],
+            "identifier": device["identifier"],
+        }
+
+        category = device["category"]
+        if category not in ["BIOS", "USB"]:
+            category = category.capitalize()
+
+        if category not in hardware_details:
+            hardware_details[category] = []
+
+        hardware_details[category].append(device_info)
+
+    # default to category, which contains the least specific form_factor
+    form_factor = release_details.get("form_factor", category)
+
+    return render_template(
+        "certified/hardware-details.html",
+        canonical_id=canonical_id,
+        model_name=models["model"],
+        form=models["category"],
+        form_factor=form_factor,
+        vendor=models["make"],
+        major_release=models["major_release"],
+        hardware_details=hardware_details,
+        release_details=release_details,
+    )
+
+
+def certified_model_details(canonical_id):
+
+    model_releases = api.certified_model_details(canonical_id=canonical_id)[
+        "objects"
+    ]
+
+    if not model_releases:
+        abort(404)
+
+    component_summaries = api.component_summaries(canonical_id=canonical_id)[
+        "objects"
+    ]
+
+    release_details = {"components": {}, "releases": []}
+    has_enabled_releases = False
+
+    for model_release in model_releases:
+        ubuntu_version = model_release["certified_release"]
+        arch = model_release["architecture"]
+
+        if arch == "amd64":
+            arch = "64 Bit"
+
+        release_info = {
+            "name": f"Ubuntu {ubuntu_version} {arch}",
+            "kernel": model_release["kernel_version"],
+            "bios": model_release["bios"],
+            "level": model_release["level"],
+            "notes": model_release["notes"],
+            "version": ubuntu_version,
+            "download_url": get_download_url(model_release),
```

Comment:
The logic for choosing a download url is in the template now, so I think it is safe to drop this.

Reply:
Done. Thanks.